prover/prover.go

```diff
@@ -328,6 +354,35 @@ func (p *Prover) onBlockVerified(ctx context.Context, event *bindings.TaikoL1Cli
 	}
 
 	log.Info("New verified valid block", "blockID", event.Id, "hash", common.BytesToHash(event.BlockHash[:]))
+
+	// cancel any proofs being generated for this block
+	if cancel, ok := p.currentBlocksBeingProven[event.Id.Uint64()]; ok {
+		log.Info("Cancelling proof for ", "blockID", event.Id, "hash", common.BytesToHash(event.BlockHash[:]))
+		cancel()
+		delete(p.currentBlocksBeingProven, event.Id.Uint64())
+		log.Info("Cancelled proof for ", "blockID", event.Id, "hash", common.BytesToHash(event.BlockHash[:]))
+	}
+
+	return nil
+}
+
+// onBlockProven cancels proof generation if the proof is being generated by this prover,
+// and the proof is not the oracle proof address.
+func (p *Prover) onBlockProven(ctx context.Context, event *bindings.TaikoL1ClientBlockProven) error {
+	metrics.ProverReceivedProvenBlockGauge.Update(event.Id.Int64())
+	// if oracle prover, dont cancel proof.
+	if event.Prover == p.oracleProverAddress {
+		return nil
+	}
+
+	// cancel any proofs being generated for this block
```

Comment:
Would add a check here that the proof generation is only cancelled when the correct proof (the correct fork choice) was submitted in the smart contract. So that means that `parentHash` and `parentGasUsed` are the expected values. Otherwise someone could submit the wrong proof and everybody would simply stop generating their (hopefully) correct proof which wouldn't be great. The `BlockProven` event however does not seem to contain the `parentGasUsed` (or even the `gasUsed`) values, seems like those should be added in the event for simplicity.

Reply:
I have a PR open to add getOracleProver for comparisons, I will add those to the event in the same PR!

Reply:
@Brechtpd can you take a look at the latest commits and see if this makes sense?
1) on blockProven, we check parentGasUsed + parentHash
2) on blockVerified, we dont need to check those, so we just cancel